solutions/assignment-2.2.py

```
@@ -0,0 +1,279 @@
+
+    def run(self):
+        self.download_range()
+
+    def download_range(self):
+        self.client_socket = socket.socket()
+        try:
+            self.client_socket.connect((self.host, self.port))
+        except:
+            logging.error('Could not establish connection')
+        self.client_socket.send(self.request)
+        self.receiving_data_file = tempfile.NamedTemporaryFile()
+        self.receiving_data_file_name = self.receiving_data_file.name
+        while True:
+            self.data_received = self.client_socket.recv(BUFF_SIZE)
+            if not self.data_received:
+                break
+            self.receiving_data_file.write(self.data_received)
+        self.receiving_data_file.flush()
+        try:
+            self.temp_downloaded_file = open(
+                self.receiving_data_file_name, 'r'
+            )
+        except:
+            logging.error('File connection could not be established')
+        self.i = 0
+        while self.i <= self.header_count:
+            self.temp_downloaded_file.readline()
+            self.i += 1
+        self.range_download_file.write(self.temp_downloaded_file.read())
+        self.range_download_file.flush()
+
+
+class DownloadFile(threading.Thread):
+
+    def __init__(self, host, port, path, threads, output_file):
+        self.host = host
+        self.port = port
+        self.path = path
+        self.threads = threads
+        self.output_file = output_file
+        self.request = self.path + 'HTTP/1.1\n'
+        self.request += 'Host: ' + self.host + '\n'
+        self.request += 'Range: bytes=0-' + '\n\n'
+        threading.Thread.__init__(self)
+
+    def run(self):
+        self.get_response_headers()
+        self.download_file()
+
+    def get_response_headers(self):
+        self.client_socket = socket.socket()
+        try:
+            self.client_socket.connect((self.host, self.port))
+        except:
+            logging.error('Could not establish connection')
```

You should probably reraise the exception. Ignoring it is a bad idea, since `recv` below will then error out.